lib/open3.rb

```diff
@@ -899,48 +899,75 @@ def capture2e(*cmd)
+  # that:
   #
-  # Each cmd is a string or an array.
-  # If it is an array, the elements are passed to Process.spawn.
+  # - Creates a child process for each of the given +cmds+
+  #   by calling Process.spawn.
+  # - Pipes the +stdout+ from each child to the +stdin+ of the next child,
+  #   or, for the first child, from the caller's +stdin+,
+  #   or, for the last child, to the caller's +stdout+.
+  # - Waits for all child processes to exit.
   #
-  #   cmd:
-  #     commandline                              command line string which is passed to a shell
-  #     [env, commandline, opts]                 command line string which is passed to a shell
-  #     [env, cmdname, arg1, ..., opts]          command name and one or more arguments (no shell)
-  #     [env, [cmdname, argv0], arg1, ..., opts] command name and arguments including argv[0] (no shell)
+  # With no block given, returns a 3-element array containing:
   #
-  #   Note that env and opts are optional, as for Process.spawn.
   #
-  # The options to pass to Process.spawn are constructed by merging
-  # +opts+, the last hash element of the array, and
-  # specifications for the pipes between each of the commands.
+  # - The +stdin+ stream of the first child process.
+  # - The +stdout+ stream of the last child process.
+  # - An array of the wait threads for all of the child processes.
   #
   # Example:
   #
-  #   Open3.pipeline_rw("tr -dc A-Za-z", "wc -c") {|i, o, ts|
-  #     i.puts "All persons more than a mile high to leave the court."
-  #     i.close
-  #     p o.gets #=> "42\n"
-  #   }
-  #
-  #   Open3.pipeline_rw("sort", "cat -n") {|stdin, stdout, wait_thrs|
-  #     stdin.puts "foo"
-  #     stdin.puts "bar"
-  #     stdin.puts "baz"
-  #     stdin.close     # send EOF to sort.
-  #     p stdout.read   #=> "     1\tbar\n     2\tbaz\n     3\tfoo\n"
-  #   }
+  #   Open3.pipeline_rw("sort", "cat -n")
+  #   # => [#<IO:fd 8>, #<IO:fd 9>, [#<Process::Waiter:0x000055f1d78dd7d0 sleep>, #<Process::Waiter:0x000055f1d78dd438 run>]]
+  #
+  # With a block given, calls the block with the +stdin+ stream of the first child,
+  # the +stdout+ stream  of the last child,
+  # and an array of the wait processes:
+  #
+  #   Open3.pipeline_rw("sort", "cat -n") do |first_stdin, last_stdout, wait_threads|
+  #     first_stdin.puts "foo"
+  #     first_stdin.puts "bar"
+  #     first_stdin.puts "baz"
+  #     first_stdin.close # send EOF to sort.
+  #     puts last_stdout.read
+  #     p wait_threads
+  #   end
+  #
+  # Output:
+  #
+  #   1	bar
+  #   2	baz
+  #   3	foo
+  #   [#<Process::Waiter:0x000055f57082bc00 dead>, #<Process::Waiter:0x000055f57082b7c8 dead>]
+  #
+  # Like Process.spawn, this method has potential security vulnerabilities
+  # if called with untrusted input;
+  # see {Command Injection}[rdoc-ref:command_injection.rdoc].
+  #
+  # Unlike Process.spawn, this method waits for the child processes to exit
+  # before returning, so the caller need not do so.
```

Comment:
Ditto, I don't think this is true. The caller should have to wait using the `wait_threads` returned.

Reply:
Fixed.